docs/design/2020-04-17-sql-sequence.md

```diff
@@ -0,0 +1,115 @@
+
+## Background
+
+The columns in TiDB with 'auto-increment' attribute will cache batch values by default in each instance to ensure better performance, which means auto-increment only guarantees the uniqueness under a multi-tidb-node cluster. This kind of cache mode is uncontrollable and implicit in auto-increment.
+
+But in sequence, we take it open. Users can specify whether to use cache mode or not, what the cache size should be, what's the min and max of the range like and so on. This factors make sequence more flexible, controllable and easy to use.
+
+## Syntax and Semantics
+
+Create sequence statement:
+
+```
+CREATE SEQUENCE [IF NOT EXISTS] sequence_name
+[ INCREMENT [ BY | = ] increment ]
+[ MINVALUE [=] minvalue | NO MINVALUE | NOMINVALUE ]
+[ MAXVALUE [=] maxvalue | NO MAXVALUE | NOMAXVALUE ]
+[ START [ WITH | = ] start ] 
+[ CACHE [=] cache | NOCACHE | NO CACHE ]
+[ CYCLE | NOCYCLE | NO CYCLE]
+```
+ 
+Show create sequence statement:
+
+```
+SHOW CREATE SEQUENCE sequence_name
+```
+
+Drop sequence statement:
+
+```
+DROP SEQUENCE [IF EXISTS] sequence_name_list
+```
+
+Get the next value of sequence:
+
+```
+NEXT VALUE FOR sequence_name / NEXTVAL(sequence_name)
+```
+
+Get the previous value of sequence:
+
+```
+LASTVAL(sequence_name)
+```
+
+Move the current position in sequence number axis:
+
+```
+SETVAL(sequence_name, num)
+```
+
+## Rationale
+
+### Sequence metadata
+
+Sequence is a table-level object, sharing the same namespace with the base table and view objects, whose meta are recorded in `*model.tableInfo`. But actually sequence definition do not have columns in its meta, which means the sequence value cannot be operated by `DML` interface. The only way to do it is by `setval` function, after all, the sequence value is essentially a KV pair with some constraints.
+
+### Sequence values
+
+Each sequence value will be stored in its own range, with the key format below:
+
+```
+DB:<dbID> SID:<sequenceID> ——> int64 
+```
+
+Reads and writes to the sequence value (via the functions `nextval`, `setval`), will be implemented by direct calls to the `KV` layer's `Get`, `Inc`, and `Set` functions. Since sequence objects store the value at the range of its own, it will cause hotspot at sequence meta under insertion intensive cases.
+
+### Sequence Transaction
+
+Storing sequence values in their own ranges means that insertions to tables which use sequences will always touch two ranges. However, since the sequence update takes place outside of the SQL transaction, this should not trigger the `2PC` commit protocol. 
+
+Sequence computation is under the expression framework, evaluating one value at a time. The values fetched by these functions won't be rolled back when the outer SQL transaction is aborted.
```

Comment:
What's the meaning of `the values ... won't be rolled back ...` ?

Reply:
It means that the sequences that have been allocated will not be 'returned back' to the system, they'will be abandoned with the transaction.

Reply:
yes + 1:
```
create sequence seq;
begin
insert into t values nextval();
abort
the nextval will be 2, 1 is used in aborted txn